hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/QuasiClosedStuckUnderReplicationHandler.java

```diff
@@ -0,0 +1,158 @@
+    this.currentContainerSize = (long) conf.getStorageSize(ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE,
+        ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE_DEFAULT, StorageUnit.BYTES);
+    this.replicationManager = replicationManager;
+    this.metrics = replicationManager.getMetrics();
+  }
+
+  @Override
+  public int processAndSendCommands(Set<ContainerReplica> replicas, List<ContainerReplicaOp> pendingOps,
+      ContainerHealthResult result, int remainingMaintenanceRedundancy) throws IOException {
+    ContainerInfo containerInfo = result.getContainerInfo();
+    LOG.debug("Handling under replicated QuasiClosed Stuck Ratis container {}", containerInfo);
+
+    int pendingAdd = 0;
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.ADD) {
+        pendingAdd++;
+      }
+    }
+
+    if (pendingAdd > 0) {
+      LOG.debug("Container {} has pending add operations. No more replication will be scheduled until they complete",
+          containerInfo);
+      return 0;
+    }
+
+    QuasiClosedStuckReplicaCount replicaCount =
+        new QuasiClosedStuckReplicaCount(replicas, remainingMaintenanceRedundancy);
+
+    List<QuasiClosedStuckReplicaCount.MisReplicatedOrigin> misReplicatedOrigins
+        = replicaCount.getUnderReplicatedReplicas();
+
+    if (misReplicatedOrigins.isEmpty()) {
+      LOG.debug("Container {} is not under replicated", containerInfo);
+      return 0;
+    }
+
+    // Schedule Replicas for the under replicated origins.
+    int totalRequiredReplicas = 0;
+    int totalCommandsSent = 0;
+    IOException firstException = null;
+    List<ContainerReplicaOp> mutablePendingOps = new ArrayList<>(pendingOps);
+    for (QuasiClosedStuckReplicaCount.MisReplicatedOrigin origin : misReplicatedOrigins) {
+      totalRequiredReplicas += origin.getReplicaDelta();
+      List<DatanodeDetails> targets;
+      try {
+        targets = getTargets(containerInfo, replicas, origin.getReplicaDelta(), mutablePendingOps);
+      } catch (SCMException e) {
+        if (firstException == null) {
+          firstException = e;
+        }
+        LOG.warn("Cannot replicate container {} because no suitable targets were found.", containerInfo, e);
+        continue;
+      }
+
+      List<DatanodeDetails> sourceDatanodes = origin.getSources().stream()
+          .map(ContainerReplica::getDatanodeDetails)
+          .collect(Collectors.toList());
+      for (DatanodeDetails target : targets) {
+        try {
+          replicationManager.sendThrottledReplicationCommand(containerInfo, sourceDatanodes, target, 0);
+          // Add the pending op, so we exclude the node for subsequent origins
+          mutablePendingOps.add(ContainerReplicaOp.create(ContainerReplicaOp.PendingOpType.ADD, target, 0));
+          totalCommandsSent++;
+        } catch (CommandTargetOverloadedException e) {
+          LOG.warn("Cannot replicate container {} because target {} is overloaded.", containerInfo, target);
```

Comment:
It's the sources that are overloaded here, as the push replicate commands are being sent to them. So instead of `target` we need to log `sourceDatanodes`.

Reply:
Ah, well spotted. I keep getting source and target confused! I will fix this.